### DISCUSSION TITLE
Where/how to install user scripts and tools

### DISCUSSION THREAD

**Discussion by jonct** [Jip-Hop/jailmaker#239] [Q&A] [answered]
In principle, the [standard location](https://www.freedesktop.org/software/systemd/man/latest/file-hierarchy.html#Home%20Directory) for user-installed scripts should be in `~/.local/bin`. But that [breaks down](https://github.com/Jip-Hop/jailmaker/pull/238#issuecomment-2232960297) when the tool sometimes needs to be launched under `sudo`.

Has anyone else solved this well? Well enough to document for other users, such that it covers general cases?

**Comment by Lockszmith-GH** [comment]
The way I see it, there are 2 problems:
1. Privileged invocation from $PATH when in need of `sudo` - which breaks the PATH.
2. Persistence that is not tied to the user invoking the script, but to the system.

# Issue 1: Privileged invocation
We either need (a) a `sudo` invocation script, (b) a shell alias/function, or (c) have the script to invoke `sudo` for us.

* (a) Create an additional script in the user's `$PATH`, it will invoke main script with the proper `sudo` arguments.  
   Pros: main script does not contain any privilege escalation code.  
   The main issue the needs to solve is the deployment of this script.  
   A possible solution would be to have the main script generate this simple shell script in any location we choose (and it's the user's responsiblity to make sure it's in the `$PATH`)
* (b) shell alias/function:  
   Similar to what we have today in the `README`. 
   Users will need to edit their `.bashrc` files for this
* (c) escalate privileges from the script  
  In order for this not happening automatically, we can require a `--with-sudo` argument that will invoke the script with the same arguments under `sudo`.

I like the user-side simplicity of (a), for the dev side, (c) seems rather simple.

# Issue 2: Persistence 
I'll just add that ther requirement is specifically in TrueNAS SCALE - which gives us OS access (including root), but only an immutable read-only OS root file-system.
While this is a problem, we also benefit from knowing the OS is TrueNAS SCALE which comes with some 'perks', like ZFS.

I say, let's have the script do a double-take autodetection and installation - similar to what it does today, utilizing ZFS User Properties.

We can still use environment or cli arguments, but use a similar approach to iX, by storing metadata in a location we designate, and then rely on that data exclusively.

As long as we set the properties at the root of the pool, I believe we shouldn't hit any performance snags.
The path in the props should be relative to the root of the pool its set on.

Example:
* path of jails root is: `/mnt/tank/mydata/myjails`
* prop should be: `mydata/myjails`

So, order of detection:

1. a cli parameter
4. env variable
5. auto detect via ZFS props

As long as there is **only 1** such pool, we should be good to go.
If you think multiple pools might exist, you can choose to add an `active-on-hostname` or just `active` prop, but that's a design descision I'm not sure we need to make now.

Example on setting the prop:

```console
# zfs set "com.github.jip-hop.jailmaker:jails_path=mydata/myjails" tank
```

reading the props (`d0` here makes sure we're only querying the root of all the attached pools):

```console
# zfs get -Hd0 -s local "com.github.jip-hop.jailmaker:jails_path"
tank       com.github.jip-hop.jailmaker:jails_path       mydata/myjails  local
```

Followed by:

```console
# sudo zfs get -Hd0 "mountpoint" tank/mydata/myjails
tank/mydata/myjails mountpoint      /mnt/tank/mydata/myjails    default
```

**Reply by jonct** [reply]
> (c) escalate privileges from the script

This appeals to me. I suppose there's also an option (d) to document installation *only* for root, or (e) once for user *and* once for root. But I do think it's important for `jlmkr` not to overreach for benign actions like help, list, status, etc.

> In order for this not happening automatically, we can require a `--with-sudo` argument that will invoke the script with the same arguments under `sudo`.

This eases my mind. I had considered an internal rebound launcher for just the few actions that need it, but having the user explicitly authorize this by argument is a good practice. (Having them respond to *unexpected* password prompts is a bad practice. Having sudo unexpectedly hold the door open for another five minutes is another. Do we dare guide users into `visudo`?)

> we also benefit from knowing the OS is TrueNAS SCALE

Timely point. We don't know the user's choice of shell up front, but we do know the lay of the land. We *could* start advocating for something like `/home/.shared/bin` but I'd want to run that by the iX braintrust first.

Also since users are running `jlmkr startup` as a Post-Init Script we could *\[with permission\]* do something at boot time. Whether that's making sure /root has the goods, or even overlay-mounting jlmkr straight into `/usr/local/bin` at boot time…

Gotta say it: Jip-Hop had certainly solved where to park the tool by default. 😘

> As long as there is **only 1** such pool, we should be good to go. If you think multiple pools might exist, you can choose to add an `active-on-hostname` or just `active` prop, but that's a design descision I'm not sure we need to make now.

That's an important topic, but I think it's a separate one. First thing is to get the script running as root.

Once the script is running, then our code can work out where to find its cues. That feels like a [solved](https://github.com/jonct/jailmaker/blob/d2c39292de43504b4a88a833eabfc6bab431b06f/src/jlmkr/paths.py#L27-L30) or at least solvable problem.

**Reply by jonct** [reply]
The more I think about this, `jlmkr --sudo create …` feels about right. As written; no shorthand flag. It's slightly annoying, but explicable in one tiny sentence. And makes absolutely clear what's going on without any underhanded surprises.

I'd suggest that the actions that need it should exit noting how/why, and that all other actions should proceed after a warning that it wasn't needed (and won't be applied).

And if the user elects to go automatic, there's the option of a `JAILMAKER_AUTO_SUDO=1` and/or an `escalate_privilege = auto`.

@Lockszmith-GH's user-friendly jump script (a) could boil down to `alias jlr="JAILMAKER_AUTO_SUDO=1 jlmkr"` so it's not a mysterious download but something the user actively sees and reads if she opts to paste it into .zshrc.

**Reply by Lockszmith-GH** [reply]
> Timely point. We don't know the user's choice of shell up front, but we do know the lay of the land. We _could_ start advocating for something like `/home/.shared/bin` but I'd want to run that by the iX braintrust first.

`/home` is on the root pool which is read-only, we can't use that for sure, so we don't have a `/home/.shared` entry either.

That's why, whatever solution we come up with, while it will work - it will have some convolution to it.

The way I see it, `heavyscript` has the correct approach.
* The installation mechanism is embedded within the script.
* The installation mechanism is relying on `git` (which is available, because it's available on TrueNAS SCALE).
* It's always under `/root` - this is the risky part... as it's on the boot-pool, specifically `boot-pool/ROOT/<Versoin>/root` unsure if it is preserved across upgrades.

This is where I think we can diverge a bit:

* Wherever the script runs it's `setup` action, it will do the following:
  1. Only work via `sudo` (for example, it will look for `$SUDO_USER` env).
  2. Install (using git) in a `~root/jailmaker` dir. While the dir might dissapear, it could be repaired by the copy/stub of the script in `~/.local/bin` if needed.
  3. Install a stub (probably just a copy of the main script) in the user's `~/.local/bin`, this will be able to perform the benign operations that don't require elevation

**Reply by jonct** [reply]
> `/home` is on the root pool which is read-only, we can't use that for sure

`/home` and `/root` are both on the boot pool, but are *not* mounted read-only. They both survive upgrades but not rebuilds.

> The way I see it, `heavyscript` has the correct approach.

The old heavyscript, or the forked-and-borked aftermath? I'll try to keep my mind open, but I don't know what git provides over curl or wget when you're not pulling down a full tree. Also I **really truly** don't want to encourage anyone to download-and-sudo in one swift motion. Nor does it seem necessary.

Do you not think this can stay in a single file, even if that file shells out to itself with escalated privileges under verifiable conditions?

**Reply by Lockszmith-GH** [reply]
> Do you not think this can stay in a single file, even if that file shells out to itself with escalated privileges under verifiable conditions?

I didn't understand your question, could you please elaborate?

**Reply by jonct** [reply]
OK… heavyscript has changed a lot since I saw it last. **Impressive** body of work, all around!

The documentation probably just accidentally lags behind current practice in how it installs itself (and that won't change; it's EOL). The [readme](https://github.com/Heavybullets8/heavy_script#how-to-install) still claims to add a symlink into `/usr/local/bin` and says nothing about `/etc/sudoers`. [What it does](https://github.com/Heavybullets8/heavy_script/blob/main/functions/deploy.sh) there is pretty sophisticated, and not obfuscated. But IMO it's also not accurately explained to its users, at least on their first encounter.

Briefly: it does pull down a fairly sizable set of files, has a lot of surface area and cloud dependencies. It self-modifies in place, and helps itself to a broad range of the system. I *do not* mean to suggest that any one of its users feels tricked or betrayed by that. But after due scrutiny, I'm hesitant to embrace it as a role model.

Credit where due: it can be installed *either with or without* root privilege, and it does marvelously explain that point right up front in the readme. Does it later require root access, at which point which it does this adaptation of `sudoers` and makes changes to its own ownership? Dunno, but from what I've seen so far I'm confident that it would dutifully prompt and explain well enough if it does.

FWIW: it does reference an `.ini` file for settings and locations of the relevant datasets. Neat!

I **am** in favor of including a built-in self-installation mechanism. In that case a short wizard actually feels like an appropriate interaction model. We'll still have the question of *where* to install, but so long as we're tweaking shell resource files we can respect a user's override of our clear recommendation.

On my system `root` uses `zsh`. Was that the default or did I go to the trouble to change it some time back? If it's generally consistent then I'd suggest we prompt for permission to append path extensions in all three of `~/.bashrc`, `~/.zshrc`, and `/root/.zshenv`.

I can't get to this right now, but I'd be up to taking point on something like…
```
$ ./jlmkr self-install
INSTALLING JAILMAKER 3.0.0.dev2

  ▶️ The jlmkr tool is not yet installed.

INSTALL FOR USER [step 1 of 2]

Your current search path looks for scripts in the following directories,
in order. You can extend this list by editing your your ~/.zshrc file.
Some directories are reserved by the system and cannot be changed.

  🆗 /mnt/pool/users/jonct/.local/bin
  🚫 /usr/local/bin
  🚫 /usr/bin
  🚫 /bin
  🚫 /usr/games

Install to /mnt/pool/users/jonct/.local/bin? [y/n/e to edit]: n

Install to /home/.shared/bin? [y/n/e to edit]: y

  ✅ Installed at /home/.shared/bin/jlmkr

INSTALL FOR SUDO [step 2 of 2]

Some operations require elevated privileges. The root user
must also have this directory in its own search path.

Ensure that it does? [y/n]: y
[sudo] password for jonct:

  ✅ Installed

$
```
… where it might detect an existing current or older version, might walk a sequence of available directories, might reach the end with the user noping out, etc.

**Reply by jonct** [reply]
> > Do you not think this can stay in a single file, even if that file shells out to itself with escalated privileges under verifiable conditions?
> 
> I didn't understand your question, could you please elaborate?

I see the user downloading a .zip file from the Releases page. From there it's just a standalone tool that knows how to prompt to escalate its own privileges when necessary. Whether that's to invoke a big elaborate wizard or just to complete the task the user requested. But it came from a download, and launched like a script. No git, no sync, no magic. 😌

**Reply by Lockszmith-GH** [reply]
I personally hate the 'download from release' + 'extract' + 'change permissions'.
I'm ok with having it work like that, but I would like a single `curl` piped to `sh` that does the work properly (including validating has on download).

that's why utilizing git makes sense to me. Yes, it will pull unnecessary files, but it will also have a simple auto-update mechanism: `git fetch`, followed by `git pull`. `git ls-remote origin` can provide a detection mechanism.
But... This all works if we keep it in script form. If we switch to binary - then, a single-curl-to... (sh/python) should do the work.

I really like the way **webi installer** [🔗](https://webinstall.dev) is designed in that aspect.

I think we can draw from it in designing the installer logic

**Reply by jonct** [reply]
Permissions repair is an unfortunate side-effect of doing it your way. That part doesn't even come up under mine.

    curl -L https://github.com/Jip-Hop/jailmaker/archive/refs/tags/jlmkr-3.0.0.dev999.tar.gz | tar -xz jlmkr
    ./jlmkr self-install

I can't/won't pipe remote content to shell. That's a big red flag for me. Under extreme circumstances I'll download, read, and follow along. But I'd be constantly suspicious of what other casual overreach might be waiting for me on the inside.

I have no objections at all to a `jlmkr self-update`. That's a helpful *optional* convenience for many.

**Reply by Lockszmith-GH** [reply]
I agree with the sentiment, but that's what signing is for.
I mean - I either trust a package completely - after I vet it - or I don't.

If the installer is part of the pacakge - I vet it before I trust it. I see no difference between `./jlmkr self-install` and running something directly from the web.

But I agree, that whatever is done, both options should be available.

**Reply by jonct** [reply]
> I mean - I either trust a package completely - after I vet it - or I don't.

Spot on. Jip-Hop has gotten iX's attention; with luck and more hard work I hope they'll consider vetting/vouching/bundling directly.

My point here is basically that an unfamiliar project resorting to `curl | bash` is a shibboleth to me. A code smell, if you'd prefer. A subjective warning that maybe the devs are a little cavalier with other people's systems and standards. Not that it suggests any impure motives, but that due care wasn't always taken and corners have been cut — both up front and throughout the package.

Some projects are so small and unscrutinized that they have no excuse getting away with that at all. Jailmaker has earned a substantial and growing audience; maybe it's an exception. If not, it's getting there. Maybe it eventually gets coronated by iX, in whole or in part. My personal take is that its growth is to some degree, rightly *and* wrongly, gated by this first impression.

> I see no difference between ./jlmkr self-install and running something directly from the web.

Indeed. So *for me*, I just want to move that tool in with my others and be done with it. Maybe add a config file somewhere sane.

I recognize that some users may need more hand-holding than that. And that too few of us share the responsibility to provide it. So whether they'd prefer a magic copy-paste block, or an approachable readme with links to in-depth adjunct material, or a single typeable command, or any combination of these… all that I feel strongly about is that they can see in advance exactly what it'll do.

> But I agree, that whatever is done, both options should be available.

I think we agree on why. 🤝 I'm grateful we're having a productive discussion on how.

**Reply by Lockszmith-GH** [reply]
Actually, what I look for in these 'one-line installers' is automation, not hand-holding.
I'm interested in having the ability to script a system into existance.
That's my drive. It's not about hand-holding.
I don't want the script to do anything other than place it in an exceptable location, and if some tying in to the system is required, to state clearly what that is, and I'll do it.

What I see in the scope of an installation script:
- Placing the necessary exceutables in a the correct location (`/root/bin` for example) or a pool location with good configuration options to point at (like the zfs props, or env variable).
- Being able to fix any missing component in case iX reverts the boot-pool
- Allowing for an easy update detection and automate the upgrade process to make is as simple as possible

- Elevation, only when explicitly asked for.

**Reply by jonct** [reply]
I like it! I suppose an interactive wizard isn't the [right tool](https://github.com/Jip-Hop/jailmaker/discussions/231#discussioncomment-10012873) for *this* job, either. 🧞

In effect…
- README.md
- supplemental wiki or Advanced-Detail.md
- `install-jlmkr.sh`

… are all just equivalent means/media to document the steps needed to install. All three encourage responsible usage; none leads users into any black boxes or scary roadblocks.

I think any *(or all? some?)* beats code-golfing an ornate wrapped one-liner block to be copied/pasted wrong if not quietly munged by a CMS or browser first.

> Being able to fix any missing component in case iX reverts the boot-pool

This is the only point for which I'd say my interactive installer transcript would be more effective. (I implied but didn't state that it's really a "doctor" transcript: checking and offering to fix any piece that needs help.)

But we can make up the difference with clear, meaningful error messages at runtime.

> Elevation, only when explicitly asked for.

Ideally: only when explicitly **granted** at launch. Don't just leave the chainsaw on the table; don't *be handed a chainsaw.* 🌋

We still haven't figured that out yet, unless we're comfortable with `/home/.shared/bin` or similar. (`/root/bin` isn't user-accessible; user and root would end up with different versions installed. As would other users.) Even keeping the tool at the root of its own fiefdom doesn't solve this; that's only worked due to promoting reflexive sudo.

I don't yet quite follow why you'd want automated updates if you don't want automated installation, but I'll go along. Some helpful precedent for that might be the [k9scli](https://k9scli.io) project: it tells you when there's a new version, but doesn't embed any magic code to go get and install it. You simply *\[re-\]* download the tool, and put it where you want it.

**Reply by Lockszmith-GH** [reply]
> unless we're comfortable with `/home/.shared/bin` or similar. (`/root/bin` isn't user-accessible;

I'm not sure `/home/.shared` is viable, but `/mnt/tank/<somewhere>` is, and we can create symlinks from `~root/bin` as well as `~<our_user>/bin`

**Reply by Lockszmith-GH** [reply]
> I don't yet quite follow why you'd want automated updates if you don't want automated installation

I want automated everything, as long as I can take a part that process and make it my own.
Every update process has the following steps:
1. check if an update exists
2. download the necessary files
3. perform the update

As long as all of these steps can be achieved by a single tool, and I can perform them to my liking, I want it.
We don't have a package manager to rely on, so I say we roll our own, overly-simplified process into the script. 

My main gripe of not having that handy is context switching.

You ask: "How hard is it to open the web browser, copy the URL from the release page, download the file with `curl`, extract it, and let's not forget to remove any flufh from the previous version?"

(Please read everything in a friendly, jokingly voice, NOT to be read in a passive-agressiv tone)
It's not hard, but if it can be automated, I want that option.

> I mean, "How hard is it to actually use systemd-nspawnd without jailmaker?"  
> Well, we're here - aren't we?

**Reply by jonct** [reply]
Ah — sorry; I caught your edit late.

Yeah, I'm totally cool with your opening a web browser, copying an install script's URL or a magic shell incantation, and proceeding from a downloaded script rather than from a downloaded archive. For the record I have no qualms if *you choose* to pipe the script directly to bash. (I do hope that this last part is never seen as an official recommendation.)

Overall I think we're on the same page on posting a clean, simple, sing-along installer script. Putting all installation steps in a well-documented script frees up the readme to instead focus on requirements and basic operation. What's left would read more like it's intended for human consumption.

I currently have a *slight* preference that any installer/updater script not be buried inside the tool itself but alongside. Self-modifying root-privileged code has its own reputation.

But a nice readable shell script, easily found and summoned from the top of the repo, would be great. And I'm sure some user somewhere would even feed it to cron, for silent updates.

I see this as entirely compatible with a downloadable release archive, containing a tool and a readme. I see the script pulling from the archive, and I see checksums being based on the archive. (Call me a refugee of DOS-to-Mac-to-Unix text file exchange: I suppose I see software as needing differentiated, pure binary handling.)

Prioritizing scriptability seems to only prefer .tar.gz over .zip. And that makes me happy.

**Comment by jonct** [comment]
> Has anyone else solved this well? Well enough to document for other users, such that it covers general cases?

OK… I've pored through the latest `sudo(1)`, `sudoers(5)`, `sudo.conf(5)`, and `visudo(8)` to refresh my memory.

### Recommendation

1. Create a `/root/bin` directory.
2. Edit `sudoers` to tack on `/root/bin` to the *end of* the `secure_path` default.
3. Symlink individual authorized tools into `/root/bin`.

### How to

    sudo mkdir /root/bin

Always edit `sudoers` using the supplied guard rails.

    sudo visudo

The relevant line to edit is fairly clear.

    Defaults secure_path = /sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin

Extend the path with `:/root/bin` at the end, so that built-in/protected tools always still take precedence over user-added tools, particularly when called by other scripts.

Then, for example, if you install `~/.local/bin/jlmkr` for yourself, you can create a symlink. Your shell will expand the tilde before upgrading to sudo and creating the link.

    sudo ln -s ~/.local/bin/jlmkr /root/bin/

Done and done. 🥳

### Also considered

Trying to chart a decision tree through all the permutations of `/root/.bashrc` and `/root/.zshenv` and such could be a mess.

`/root/bin` on its own is not visible to users *outside* of sudo.

I don't think there's a good/responsible way to explain how to use `visudo` to whitelist a particular user to a particular script. Nor do I see a way (through the above components) to alias a tool name to an absolute path, so it's a moot point.

The default `secure_path` setting is there for good reason and should not be bypassed.

Having `secure_path` point to a regular user-managed directory is calling for trouble. (Though not *too* bad, provided you ensure the appropriate precedence.)

But Debian's choice of `/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin` intends for us to add items to `/usr/local`, and iX has that walled off from us. So IMHO we should append `/root/bin` (or `/root/.local/bin`) to the *end of* sudo's built-in paths, and we'll stay in line with Debian's security posture. Then the user can curate individual authorized tools through symlinks to wherever they live.

### Hit me

Have I missed anything?

**Reply by Jip-Hop** [reply]
Aren't edits don't with visudo lost after an upgrade of SCALE? This thread seems to indicate so but I'm not sure it applies to SCALE: https://www.truenas.com/community/threads/constant-sudo-password-prompts.9830 :

If you'd want to try you can install an older release in a VM and upgrade. Or use a nightly build as it has frequent updates available hehe.

**Reply by Jip-Hop** [reply]
According to this post it doesn't even survive a reboot :(

https://www.truenas.com/community/threads/sudo-permission-with-no-password.100625/

**Reply by Jip-Hop** [reply]
It would probably require restoring from a Pre Init script each boot: https://www.truenas.com/community/threads/persist-changes-in-etc-local-sudoers-for-remote-shutdown-via-ssh.91025/

**Reply by Lockszmith-GH** [reply]
My experience shows that anything on the boot-pool is fair-game for iX. Most of it reverts on reboot, some only revert on upgrade.

**Reply by jonct** [reply]
Oh wow — believe it or not, that's a huge advantage! Bear with me here.

The danger in editing sudoers (and the reason for visudo) would be screwing up access so badly that you can't go back and fix it. ***But*** if all it takes to fix a mistake is to remove the pre-init script and restart, then we're totally golden!

That pre-init script, I guess, would be

    sed -ie 's#Defaults secure_path = .*#&:/root/bin#' /etc/sudoers

Easy peasy?

**Reply by jonct** [reply]
> ```
> sed -ie 's#Defaults secure_path = .*#&:/root/bin#' /etc/sudoers
> ```

After a couple of tweaks and switching to **Post-Init**, this exact command does the trick for me at startup. 😃

**Reply by mrstux** [reply]
I actually recommend people clone the repo into the jailmaker dataset. 

Is that a code smell ;)

**Reply by Jip-Hop** [reply]
And what if someone edits sudo user settings in the Web GUI? I guess that will recreate the sudoers file, undo the Pre Init changes and break `jlmkr` until a reboot. It's all way to hacky if you ask me. Then I rather temporarily disable the rootfs readonly protection, create a symlink and re-enable the readonly protection (if that's possible).

But more importantly I think `jlmkr` currently makes the right compromise given the systems restrictions. Anything else will complicate the setup. Just run it with the absolute path, as `./jlmkr.py` from within the directory or through some alias.

Only iX can really solve this "Where/how to install user scripts and tools" properly (or if they accept a PR).

**Reply by jonct** [reply]
Heh. 😜 IMHO it's a perfectly cromulent administrative decision.

There could be reasonable concerns if that's the *only way it can* function… 🤔

**Reply by jonct** [reply]
> And what if someone edits sudo user settings in the Web GUI?

Huh?

**Reply by Jip-Hop** [reply]
![IMG_9540](https://github.com/user-attachments/assets/c4b41612-502a-4bc9-82e8-f0180ad578d1)

Does this not update/overwrite the sudoers file?

**Reply by Jip-Hop** [reply]
I'd say create a PR for iX with this change: `Defaults secure_path = /sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/root/bin` and see what happens. 😜

**Reply by jonct** [reply]
Is it possible that you're working a different problem than I am?

**Reply by Jip-Hop** [reply]
I don't think so. I'm just saying that your suggestion will be wiped, more often than you realize. Open a shell. Run `sudo visudo`. Make some changes. Then go to https://truenas.scale/ui/credentials/users and edit a user. Toggle `Allow all sudo commands` and hit Save. Check again with `sudo visudo`, your edits will be gone.

**Reply by jonct** [reply]
I don't imagine I'll have made such administrative changes by accident. Nor on a lark. I'm confident I'd face symptoms soon enough to recognize their connection. And in general I'd venture that one's frequency of restarts would tend to scale with the intensity of one's poking around under the hood.

Maybe the terrible horrible worst *likely* case would be upon creating a new administrative user? I could see that one. Especially when paired with one of my blindspots: I'm not really thinking in terms of mission-critical TrueNAS SCALE deployments with more than one administrator (let alone floating administrators). I tend to think the target audience is mostly in homelabbing territory.

I could offer you a technical response, in case we are truly working the title investigation at face value. Or I can offer you a social/diplomatic response if there's more to it than that. I'm honestly not sure which way to go here.

#### The technical matter

IMHO this tweak is not nearly as elegant, but is [pretty clever](https://www.reddit.com/r/freenas/comments/li7be0/comment/hz81vr5/) in its own right. The Redditor's details don't fit SCALE but his core idea does.

`/etc/sudo.conf` does exist — as a fully commented-out placeholder — and does survive a restart. (Yes, yes, not an upgrade.)

So through `sudo.conf` you can tell sudo to use a bypass policy file. That file would have two lines: 1) import the official managed sudoers file and 2) reactivate Debian's `/etc/sudoers.d` directory. Then put your customizations in the directory.

#### The diplomatic matter

I seem to be walking into a number of rakes, and… well, I'll stand by my earlier guess: making this to some degree less fun for you.

I'm not looking for a fight on any of it. Certainly not on this topic, which is ultimately not even about Jailmaker. It just emerged from Jailmaker discussions, and showed promise for generalization to a longstanding problem that a number of us have identified.

I don't know how to offer you some space without risk of it coming across as a petulant threat to walk away from the project. *(That is, unless my going away would allow you to stay, in which case we **unanimously** prefer that you do!)* But I do happen to have quite a bit to catch up on after this week, so it's sincere: I'm *not disengaging* but I'll go attend to some other priorities for a bit.

Much affection, my friend. ✌️

**Reply by Jip-Hop** [reply]
Good job on the research! You now have all the ingredients to make a solution which will survive reboot, upgrade and `sudo` related edits made through the GUI. 😀 And great to see [you've opened a feature request about this on the forum](https://forums.truenas.com/t/decide-and-document-where-to-install-tools-and-scripts/9388). Perhaps post your idea there too? There's a better chance iX will chime in there.

All the best and wish you luck with the catching-up.

**Comment by Jip-Hop** [comment, marked as answer]
What do you think of this as a general solution?

Post Init Command:

```bash
mkdir -p /root/usr-local-bin && mount -o remount,rw /usr && mount -t overlay overlay -o lowerdir=/root/usr-local-bin:/usr/local/bin /usr/local/bin; mount -o remount,ro /usr
```

This will temporarily remount `/usr` as read-write, then [overlay mount](https://wiki.archlinux.org/title/Overlay_filesystem) `/root/usr-local-bin` underneath `/usr/local/bin` and remount `/usr` back as readonly as if nothing happend. Put your executables inside `/root/usr-local-bin` and they shall be available globally for any user, including for use with `sudo`.

Since `/usr/local/bin` is the last dir in the default `secure_path` it can't override any of the commands in directories that come before it. Since `/root/usr-local-bin` is the lowest lowerdir, it can't override anything already in `/usr/local/bin` (only augment it). So looks secure to me and free of conflicts. And of course only `root` can put files in `/root/usr-local-bin`.

This will survive reboots, updates and upgrades.

I find this a more direct way to make executables globally available then by changing `sudo.conf` and related files. Of course neither way is supported by iX. But I expect the overlay mount workaround to break in less unexpected ways in the future.

It's quite elegant if you ask me. 🙂 But I agree there needs to be an officially documented (in the TrueNAS docs) way to do these kinds of things [as you suggest](https://forums.truenas.com/t/decide-and-document-where-to-install-tools-and-scripts/9388).

**Reply by jonct** [reply]
> It's quite elegant if you ask me. 🙂

Oh, ***hell yes.*** 🤩 I'd briefly pondered overlayfs but never reached the insight to sink it *underneath* `/usr/local/bin`. Brilliant!

I could bikeshed and code golf a bit, but only in the collaborative improv spirit of "yes-and"… *\[or hell-yes-and?\]*
- you can leave mkdir to the user since they populate it anyway
- cover all of `/usr/local` including `sbin` and `share/man`
- maybe `/root/.local`?

And if it made Lockszmith-GH happy, Post-Init could be *either* a one-liner or an internally well-documented `/root/.local/sbin/userusr` script (or even some name that *doesn't* suck…)

> I agree there needs to be an officially documented (in the TrueNAS docs) way to do these kinds of things [as you suggest](https://forums.truenas.com/t/decide-and-document-where-to-install-tools-and-scripts/9388).

Kris [made me](https://ixsystems.atlassian.net/browse/NAS-130029?focusedCommentId=269127) do it. 😉

**Reply by Jip-Hop** [reply]
Yeah the `mkdir -p` is optional. I did update the command to replace the last `&&` with `;` to ensure you don't end up with a read-write `/usr` in case the overlay mount fails.

**Reply by Jip-Hop** [reply]
I suddenly remember having read about a `systemd` tool which does something similar: `systemd-sysext`.

https://manpages.debian.org/bookworm/systemd/systemd-sysext.8.en.html

What I suggested appears to be a poor-man's `systemd-sysext`.

**Reply by Jip-Hop** [reply]
Turns out the Post Init Command even works this way (without first remounting `/usr` as read-write):

```bash
mount -t overlay overlay -o lowerdir=/root/usr-local-bin:/usr/local/bin /usr/local/bin
```

**Reply by jonct** [reply]
Sweet! That removes the one known element of danger, *and* makes it a clean readable one-liner. 🙌

I haven't been able to test this, yet. Does it not require `-o remount`?

For anyone experimenting: if this were applied for more than …bin then it would look like the following. And then build systems like cmake would correctly handle any needed dylibs and manfiles and such — you'd just tell them to use a prefix of `/root/.local`. (System libs will take precedence at runtime, so do link with the system's libs at build time. That's typically the default.)

    mount -t overlay overlay -o lowerdir=/root/.local:/usr/local /usr/local

Is [this note](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#changes-to-underlying-filesystems) anything of concern? My off-the-cuff guess says that's primarily about open files in a *read-write* top directory. But it *could* possibly suggest a need for some sort of flush or remount after making changes on the lower deck.

**Reply by Jip-Hop** [reply]
`/usr/local/bin` is not a mount by default so I that's why I think it doesn't need the `remount` options. The flush you mention is something to try indeed.

**Reply by Lockszmith-GH** [reply]
that's the cleanest solution, I like it very much.

This is exactly what [`jlmkr satrtup`](https://github.com/Jip-Hop/jailmaker?tab=readme-ov-file#startup-jails-on-boot) should do.

**Reply by jonct** [reply]
I may have steered you wrong on dropping mkdir. 😔 The worst case mistake wouldn't be trying to mount a directory that doesn't yet exist, but rather trying to mount one with the wrong permissions.

Casual `sudo mkdir` for me created the directory as 0700, causing iX's tools to lose visibility after restart. I'm not sure everyone would recognize what happened or what to do about it. So now I'm here:

    mkdir -pm 755 /root/.local && mount -t overlay overlay -o lowerdir=/root/.local:/usr/local /usr/local
